components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIProviderImpl.java

```diff
@@ -3436,52 +3427,92 @@ public void deleteAPI(API api) throws APIManagementException {
-                        }
+        if (apiId != -1) {
+            try {
+                cleanUpPendingAPIStateChangeTask(apiId);
+            } catch (WorkflowException | APIManagementException e) {
+                log.error("Error while executing API delete operation on cleanup workflow tasks", e);
+                error = true;
+            }
+        }
+        if (api.getId().toString() != null) {
+            Map<String, KeyManagerDto> tenantKeyManagers = KeyManagerHolder.getTenantKeyManagers(tenantDomain);
+            for (Map.Entry<String, KeyManagerDto> keyManagerDtoEntry : tenantKeyManagers.entrySet()) {
+                KeyManager keyManager = keyManagerDtoEntry.getValue().getKeyManager();
+                if (keyManager != null) {
+                    try {
+                        keyManager.deleteRegisteredResourceByAPIId(api.getId().toString());
+                    } catch (APIManagementException e) {
+                        log.error("Error while deleting Resource Registration for API " + api.getId().toString() +
+                                " in Key Manager " + keyManagerDtoEntry.getKey(), e);
                     }
                 }
             }
+        }
+
+        try {
             GatewayArtifactsMgtDAO.getInstance().deleteGatewayArtifacts(api.getUuid());
+        } catch (APIManagementException e) {
+            log.error("Error while executing API delete operation on gateway artifacts", e);
+            error = true;
+        }
+
+        try {
             apiPersistenceInstance.deleteAPI(new Organization(api.getOrganization()), api.getUuid());
-            APIEvent apiEvent = new APIEvent(UUID.randomUUID().toString(), System.currentTimeMillis(),
-                    APIConstants.EventType.API_DELETE.name(), tenantId, tenantDomain, api.getId().getApiName(), apiId,
-                    api.getUuid(), api.getId().getVersion(), api.getType(), api.getContext(),
-                    APIUtil.replaceEmailDomainBack(api.getId().getProviderName()),
-                    api.getStatus());
-            APIUtil.sendNotification(apiEvent, APIConstants.NotifierType.API.name());
+        } catch (APIPersistenceException e) {
+            log.error("Error while executing API delete operation on persistence instance", e);
+            error = true;
+        }
 
-            // Extracting API details for the recommendation system
-            if (recommendationEnvironment != null) {
-                RecommenderEventPublisher
-                        extractor = new RecommenderDetailsExtractor(api, tenantDomain, APIConstants.DELETE_API);
-                Thread recommendationThread = new Thread(extractor);
-                recommendationThread.start();
-            }
+        // if one of the above has failed throw an error
+        if (error) {
+            throw new APIManagementException("Error while deleting the API " + api.getUuid());
+        }
```

Comment:
Wondering whether we should still send the latter deletion events etc during an error in previous steps.